src/nvim/eval/typval.c

```diff
@@ -1799,8 +1799,9 @@ static inline void _nothing_conv_func_end(typval_T *const tv, const int copyID)
       assert((void *)&dict != (void *)&TYPVAL_ENCODE_NODICT_VAR); \
       tv_dict_unref((dict_T *)dict); \
       *((dict_T **)&dict) = NULL; \
-      if (tv != NULL) { \
-        ((typval_T *)tv)->v_lock = VAR_UNLOCKED; \
+      typval_T *tv_ = tv;  /* Avoid -Wnonnull-compare warning. #6847 */ \
```

Comment:
Using `static inline void _nothing_conv_empty_dict` like above may be cleaner and will make debugging easier if there appear some problem in this piece of code. If you do not add FUNC_ATTR_NONNULL_ALL it should not complain with a new function.